scripts/ub-check.py

```diff
@@ -0,0 +1,205 @@
+                            f'{mainfile.split(".")[0]}.GCC13.Sanitize'  ],
+        "x86_64 Alpine": [  f'{mainfile.split(".")[0]}.Clang.O0',
+                            f'{mainfile.split(".")[0]}.Clang.O2',
+                            f'{mainfile.split(".")[0]}.Clang.O3',
+                            f'{mainfile.split(".")[0]}.GCC.O0',
+                            f'{mainfile.split(".")[0]}.GCC.O2',
+                            f'{mainfile.split(".")[0]}.GCC.O3',  ],
+        "x86_64 Windows": [ f'{os.path.normpath(mainfile.split(".")[0])}.Clang.O0.exe',
+                            f'{os.path.normpath(mainfile.split(".")[0])}.Clang.O2.exe',
+                            f'{os.path.normpath(mainfile.split(".")[0])}.Clang.O3.exe',
+                            f'{os.path.normpath(mainfile.split(".")[0])}.GCC.O0.exe',
+                            f'{os.path.normpath(mainfile.split(".")[0])}.GCC.O2.exe',
+                            f'{os.path.normpath(mainfile.split(".")[0])}.GCC.O3.exe',
+                            f'{os.path.normpath(mainfile.split(".")[0])}.MSVC.O0.exe',
+                            f'{os.path.normpath(mainfile.split(".")[0])}.MSVC.O2.exe',  ],
+        "riscv64 Ubuntu": [ f'{mainfile.split(".")[0]}.Clang.O0',
+                            f'{mainfile.split(".")[0]}.Clang.O2',
+                            f'{mainfile.split(".")[0]}.Clang.O3',
+                            f'{mainfile.split(".")[0]}.GCC.O0',
+                            f'{mainfile.split(".")[0]}.GCC.O2',
+                            f'{mainfile.split(".")[0]}.GCC.O3',  ],
+        "arm64 MacOS": [    f'{mainfile.split(".")[0]}.Clang.O0',
+                            f'{mainfile.split(".")[0]}.Clang.O2',
+                            f'{mainfile.split(".")[0]}.Clang.O3',
+                            f'{mainfile.split(".")[0]}.CPP14.GCC13.O0',
+                            f'{mainfile.split(".")[0]}.CPP14.GCC13.O2',
+                            f'{mainfile.split(".")[0]}.CPP14.GCC13.O3',
+                            f'{mainfile.split(".")[0]}.GNU20.GCC13.O0',
+                            f'{mainfile.split(".")[0]}.GNU20.GCC13.O2',
+                            f'{mainfile.split(".")[0]}.GNU20.GCC13.O3',
+                            f'{mainfile.split(".")[0]}.GCC13.O0',
+                            f'{mainfile.split(".")[0]}.GCC13.O2',
+                            f'{mainfile.split(".")[0]}.GCC13.O3',  ],
+    }
+    compile_commands = compile_commands_dict[runs_on]
+    compile_products = compile_products_dict[runs_on]
+
+    return_status = {}
+    this_file_looks_odd = False
+    for compile_command, compile_product in zip(compile_commands, compile_products):
+        print(compile_command, end=' ')
+        result = subprocess.run(compile_command, shell=True)
+        if result.returncode != 0:
+            print(f'{RED}CE({result.returncode}){RESET}')
+            this_file_looks_odd = True
+            status_vector = [f'CE({result.returncode})']
```

Comment:
It may be better to use a sum type instead of a string.

Reply:
Sorry I didn't quite get it. you mean `status_vector = 'Compile OK'` here and `status_vector += 'AC'` later? In that case I can't easily determine a correct color to use for each status with `ub-check::get_color`.

Reply:
No. Sum type is an algebraic data type[^1], in practice, it is often called tagged union[^2].
In Rust, my meaning is something like this:
```rust
use std::fmt;

enum Status {
    AC(String),
    OK(String),
    CE(String)
}

impl fmt::Display for Status {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Status::AC(x) => write!(f, "GREEN AC: {}", x),
            Status::OK(x) => write!(f, "GREEN OK: {}", x),
            Status::CE(x) => write!(f, "RED CE: {}", x),
        }
    }
}
```
In Python, the thing is a bit complex. It looks like:
```python
from typing import TypeAlias
from dataclasses import dataclass

@dataclass(frozen=True)
class AC:
    info: str

@dataclass(frozen=True)
class OK:
    info: str

@dataclass(frozen=True)
class CE:
    info: str

Status: TypeAlias = AC | OK | CE

def print_status(status: Status):
    match status:
        case AC(info):
            print(f"{GREEN}AC: {info}{RESET}")
        case OK(info):
            print(f"{GREEN}OK: {info}{RESET}")
        case CE(info):
            print(f"{RED}CE: {info}{RESET}")

```

[^1]: https://en.wikipedia.org/wiki/Algebraic_data_type
[^2]: https://en.wikipedia.org/wiki/Tagged_union

Reply:
Thx, I'll look at it later